vertical-pod-autoscaler/recommender/base/circularbuffer.go

```diff
@@ -0,0 +1,62 @@
+package base
+
+// A queue with a fixed maximum capacity. Elements are inserted/removed in the
+// FIFO order. Elements are removed from the buffer when it runs out of capacity
+// and a new element is inserted.
+type CircularBuffer interface {
+	// Add a value to the end of the queue. On overflow returns true and the
+	// oldest value, which is also removed from the buffer. Otherwise returns
+	// (false, _).
+	Push(value float64) (bool, float64)
+
+	// Returns the elements in the buffer, ordered by time of insertion (oldest
+	// first).
+	Contents() []float64
+
+	// Returns a pointer to the most recently added element. The pointer can be
+	// used to modify the last element. It is only valid until the next call to
+	// Push().
+	Head() *float64
+}
+
+type CircularBufferImpl struct {
```

Comment:
All public structs should have comments. The comment should start with the struct name.